TSAN race in librdkafka: rd_avg_calc vs rd_avg_rollover

https://s3.amazonaws.com/clickhouse-test-reports/60878/37d7d80fb8950bb86f6c66ceb2f40e8a889570eb/integration_tests__tsan__[6_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel4_0.log

```
E           WARNING: ThreadSanitizer: data race (pid=1)
E             Read of size 4 at 0x7b78000ea568 by thread T982:
E               #0 rd_avg_calc build_docker/./contrib/librdkafka/src/rdavg.h:101:30 (clickhouse+0x1d5dedda) (BuildId: edd3f4e5f0cd284022b08fc33b68c99dea3d6f01)
E               #1 rd_kafka_broker_timeout_scan build_docker/./contrib/librdkafka/src/rdkafka_broker.c:880:25 (clickhouse+0x1d5dedda)
E               #2 rd_kafka_broker_ops_io_serve build_docker/./contrib/librdkafka/src/rdkafka_broker.c:3416:17 (clickhouse+0x1d5dedda)
E               #3 rd_kafka_broker_consumer_serve build_docker/./contrib/librdkafka/src/rdkafka_broker.c:4975:17 (clickhouse+0x1d5da59e) (BuildId: edd3f4e5f0cd284022b08fc33b68c99dea3d6f01)
E               #4 rd_kafka_broker_serve build_docker/./contrib/librdkafka/src/rdkafka_broker.c:5080:17 (clickhouse+0x1d5da59e)
E               #5 rd_kafka_broker_thread_main build_docker/./contrib/librdkafka/src/rdkafka_broker.c:5237:25 (clickhouse+0x1d5d3d59) (BuildId: edd3f4e5f0cd284022b08fc33b68c99dea3d6f01)
E               #6 _thrd_wrapper_function build_docker/./contrib/librdkafka/src/tinycthread.c:576:9 (clickhouse+0x1d6dbcbb) (BuildId: edd3f4e5f0cd284022b08fc33b68c99dea3d6f01)
E           
E             Previous write of size 8 at 0x7b78000ea568 by thread T983 (mutexes: read M0, write M1, write M2):
E               #0 __tsan_memset <null> (clickhouse+0x709cad4) (BuildId: edd3f4e5f0cd284022b08fc33b68c99dea3d6f01)
E               #1 rd_avg_rollover build_docker/./contrib/librdkafka/src/rdavg.h:170:2 (clickhouse+0x1d5f8c9d) (BuildId: edd3f4e5f0cd284022b08fc33b68c99dea3d6f01)
E               #2 rd_kafka_stats_emit_avg build_docker/./contrib/librdkafka/src/rdkafka.c:1354:9 (clickhouse+0x1d5f8c9d)
E               #3 rd_kafka_stats_emit_all build_docker/./contrib/librdkafka/src/rdkafka.c:1717:17 (clickhouse+0x1d5f73cb) (BuildId: edd3f4e5f0cd284022b08fc33b68c99dea3d6f01)
E               #4 rd_kafka_stats_emit_tmr_cb build_docker/./contrib/librdkafka/src/rdkafka.c:1898:2 (clickhouse+0x1d5f73cb)
E               #5 rd_kafka_timers_run build_docker/./contrib/librdkafka/src/rdkafka_timer.c:288:4 (clickhouse+0x1d6c60ca) (BuildId: edd3f4e5f0cd284022b08fc33b68c99dea3d6f01)
E               #6 rd_kafka_thread_main build_docker/./contrib/librdkafka/src/rdkafka.c:2021:3 (clickhouse+0x1d5f3129) (BuildId: edd3f4e5f0cd284022b08fc33b68c99dea3d6f01)
E               #7 _thrd_wrapper_function build_docker/./contrib/librdkafka/src/tinycthread.c:576:9 (clickhouse+0x1d6dbcbb) (BuildId: edd3f4e5f0cd284022b08fc33b68c99dea3d6f01)
```

test_system_kafka_consumers_rebalance_mv fails around once a day because of it: [link](https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgY2hlY2tfc3RhcnRfdGltZSA+PSBub3coKSAtIElOVEVSVkFMIDcgZGF5CiAgICBhbmQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyBMSUtFICdFJScKICAgIEFORCB0ZXN0X25hbWUgbGlrZSAnJXRlc3Rfc3lzdGVtX2thZmthX2NvbnN1bWVyc19yZWJhbGFuY2VfbXYlJwpPUkRFUiBCWSBjaGVja19zdGFydF90aW1lIGRlc2M=)

(Not to be confused with another race in librdkafka that was fixed 3 months ago: https://github.com/ClickHouse/ClickHouse/issues/56043)

--- a/cmake/sanitize.cmake
+++ b/cmake/sanitize.cmake
@@ -8,9 +8,6 @@ option (SANITIZE "Enable one of the code sanitizers" "")
 
 set (SAN_FLAGS "${SAN_FLAGS} -g -fno-omit-frame-pointer -DSANITIZER")
 
-# It's possible to pass an ignore list to sanitizers (-fsanitize-ignorelist). Intentionally not doing this because
-# 1. out-of-source suppressions are awkward 2. it seems ignore lists don't work after the Clang v16 upgrade (#49829)
-
 if (SANITIZE)
     if (SANITIZE STREQUAL "address")
         set (ASAN_FLAGS "-fsanitize=address -fsanitize-address-use-after-scope")

--- a/tests/tsan_ignorelist.txt
+++ b/tests/tsan_ignorelist.txt
@@ -7,4 +7,5 @@
 # https://github.com/ClickHouse/ClickHouse/issues/55629
 fun:rd_kafka_broker_set_nodename
 # https://github.com/ClickHouse/ClickHouse/issues/60443
-fun:rd_kafka_stats_emit_all
+fun:rd_avg_calc
+fun:rd_avg_rollover

--- a/tests/tsan_ignorelist.txt
+++ b/tests/tsan_ignorelist.txt
@@ -3,9 +3,13 @@
 #  [1]: https://clang.llvm.org/docs/SanitizerSpecialCaseList.html
 #  [2]: https://github.com/google/sanitizers/wiki/ThreadSanitizerSuppressions
 #
+# Caveats for generic entry "fun":
+# - does not work for __attribute__((__always_inline__))
+# - requires asterisk at the beginning *and* end for static functions
+#
 [thread]
 # https://github.com/ClickHouse/ClickHouse/issues/55629
 fun:rd_kafka_broker_set_nodename
 # https://github.com/ClickHouse/ClickHouse/issues/60443
-fun:rd_avg_calc
-fun:rd_avg_rollover
+fun:*rd_avg_calc*
+fun:*rd_avg_rollover*

--- a/tests/tsan_ignorelist.txt
+++ b/tests/tsan_ignorelist.txt
@@ -3,7 +3,7 @@
 #  [1]: https://clang.llvm.org/docs/SanitizerSpecialCaseList.html
 #  [2]: https://github.com/google/sanitizers/wiki/ThreadSanitizerSuppressions
 #
-
+[thread]
 # https://github.com/ClickHouse/ClickHouse/issues/55629
 fun:rd_kafka_broker_set_nodename
 # https://github.com/ClickHouse/ClickHouse/issues/60443

--- a/tests/ubsan_ignorelist.txt
+++ b/tests/ubsan_ignorelist.txt
@@ -6,7 +6,7 @@
 # See also [3] for all UBSan checks.
 #
 #   [3]: https://github.com/llvm-mirror/compiler-rt/blob/master/lib/ubsan/ubsan_checks.inc
-
+[undefined]
 # Some value is outside the range of representable values of type 'long' on user-provided data inside boost::geometry - ignore.
 src:*/Functions/pointInPolygon.cpp
 src:*/contrib/boost/boost/geometry/*

--- a/tests/ubsan_ignorelist.txt
+++ b/tests/ubsan_ignorelist.txt
@@ -6,6 +6,11 @@
 # See also [3] for all UBSan checks.
 #
 #   [3]: https://github.com/llvm-mirror/compiler-rt/blob/master/lib/ubsan/ubsan_checks.inc
+#
+# Caveats for generic entry "fun":
+# - does not work for __attribute__((__always_inline__))
+# - requires asterisk at the beginning *and* end for static functions
+#
 [undefined]
 # Some value is outside the range of representable values of type 'long' on user-provided data inside boost::geometry - ignore.
 src:*/Functions/pointInPolygon.cpp

